### DISCUSSION TITLE
Enable tuple_in_values in PGDialect Class by default

### DISCUSSION THREAD

**Discussion by FLIGHTLAMA** [sqlalchemy/sqlalchemy#6776] [Ideas]
Hi,
I would like to suggest to set in the PGDialect Class
`tuple_in_values = True `

At least in the docs (https://docs.sqlalchemy.org/en/14/core/operators.html and https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.tuple_)

I got the impression that with postgres as dialect it should work and add the VALUE expression to the statement.
However, this is blocked by ` tuple_in_values = False`  in the PGDialect Class

**Comment by CaselIT** [comment]
Hi,

What would be the advantage of setting it to True? It seems to control only how the `in` is rendered. https://docs.sqlalchemy.org/en/14/changelog/changelog_13.html?highlight=tuple_in_values#change-ff72be221ca8833a7265864456027d5e

Also note that you can change a dialect flag manually using `engine.dialect.tuple_in_values = True` but it's in general not recomented.

**Reply by FLIGHTLAMA** [reply]
Hi,
thanks for your fast reply.
We do your suggestion already and yes it works.

Let me try to explain my suggestion and what my reasoning is.
However, I can be totally wrong and just I overlooked something or draw a wrong conclusion.
(Something under the hood happens which then makes no difference in the query execution etc.)
If that is the case my bet then go ahead and close this :-).


In the below example **q_a** and **q_b** are real different queries for postgres, which can also have a different run time
behavior.
Especially for large number of arguments this makes a difference in execution time.
I used the example in the docs (see Tuple IN Expressions) and I was just surprised that basically q_a is generated not q_b.


```
from sqlalchemy import tuple_, Table, MetaData, Column, Text
from sqlalchemy.databases import postgres
from sqlalchemy.orm import Query

def test_show_case():
    TestTable = Table(
        'test_table',
        MetaData(),
        Column('a', Text),
        Column('b', Text),
    )

    query = Query([TestTable.c.a, TestTable.c.b]).filter(tuple_(TestTable.c.a).in_([(entry,) for entry in ["a", "b"]]))
    q_a = query.statement.compile(dialect=postgres.dialect(), compile_kwargs={"literal_binds": True})
    # SELECT
    # test_table.a, test_table.b
    # FROM test_table
    # WHERE(test_table.a) IN(('a'), ('b'))

    mod_dialect = postgres.dialect()
    mod_dialect.tuple_in_values = True
    q_b = query.statement.compile(dialect=mod_dialect, compile_kwargs={"literal_binds": True})
    # SELECT test_table.a, test_table.b
    # FROM test_table
    # WHERE (test_table.a) IN (VALUES ('a'), ('b'))
    ```

**Reply by zzzeek** [reply]
can you point us to what is the difference between these two queries as far as how they are interpreted by PostgreSQL ?

**Reply by FLIGHTLAMA** [reply]
Hi,
shortly explained the query with the VALUES create a "constant table" (https://www.postgresql.org/docs/12/queries-values.html)
where as the other query does not do it.
I tried to create an example dockerfile. I hope it is clear and I did not overlook something.

```
FROM postgres:12.4
ENV POSTGRES_HOST_AUTH_METHOD=trust
RUN mkdir -p docker-entrypoint-initdb.d && \
    echo "CREATE TABLE IF NOT EXISTS test_table (a text, b text)" > /docker-entrypoint-initdb.d/01-r.sql && \
    echo "INSERT INTO  test_table (a,b) values ('q','c'); INSERT INTO test_table (a,b) values ('s', 'e')" > /docker-entrypoint-initdb.d/02-r.sql && \
    echo "EXPLAIN ANALYSE SELECT a,b from test_table WHERE (a) IN (('a'), ('q'), ('c'))" > /docker-entrypoint-initdb.d/03-r.sql && \
    echo "EXPLAIN ANALYSE SELECT a,b from test_table WHERE (a) IN (VALUES ('a'), ('q'), ('c'))" > /docker-entrypoint-initdb.d/04-r.sql
```

If you run it you can check the outcome of the Explain Anaylse Query for the two queries (on my machine it looks like below)

```
docker build  -t testdb  .
docker run -it --rm --name db-test testdb
```
in the logs:
You see there is a difference in the query plan

```
......
server started

/usr/local/bin/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/01-r.sql
CREATE TABLE


/usr/local/bin/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/02-r.sql
INSERT 0 1
INSERT 0 1


/usr/local/bin/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/03-r.sql
                                              QUERY PLAN
-------------------------------------------------------------------------------------------------------
 Seq Scan on test_table  (cost=0.00..22.10 rows=13 width=64) (actual time=0.006..0.007 rows=1 loops=1)
   Filter: (a = ANY ('{a,q,c}'::text[]))
   Rows Removed by Filter: 1
 Planning Time: 0.243 ms
 Execution Time: 0.017 ms
(5 rows)



/usr/local/bin/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/04-r.sql
                                                     QUERY PLAN
--------------------------------------------------------------------------------------------------------------------
 Hash Semi Join  (cost=0.08..21.33 rows=13 width=64) (actual time=0.023..0.024 rows=1 loops=1)
   Hash Cond: (test_table.a = "*VALUES*".column1)
   ->  Seq Scan on test_table  (cost=0.00..18.80 rows=880 width=64) (actual time=0.002..0.003 rows=2 loops=1)
   ->  Hash  (cost=0.04..0.04 rows=3 width=32) (actual time=0.004..0.004 rows=3 loops=1)
         Buckets: 1024  Batches: 1  Memory Usage: 9kB
         ->  Values Scan on "*VALUES*"  (cost=0.00..0.04 rows=3 width=32) (actual time=0.001..0.001 rows=3 loops=1)
 Planning Time: 0.180 ms
 Execution Time: 0.046 ms
(8 rows)
```

**Reply by CaselIT** [reply]
I also use values since it usually has better performance for some of my use case, But I don't think sqlalchemy should render that by default, but give you the option to do so

This is already covered by this issue: https://github.com/sqlalchemy/sqlalchemy/issues/6289

The current version of `values()` cannot be used in an in, while it should. I have a modified version that I can use there, based on this https://github.com/sqlalchemy/sqlalchemy/wiki/PGValues 

The idea there was to have something like this work
```py
sa.select(TestTable.c.a, TestTable.c.b).filter(
    sa.tuple_(TestTable.c.a).in_(
        sa.values(TestTable.c.a).data([(entry,) for entry in ["a", "b"]]).scalar()
    )
)
```
The `.scalar()` is the new bit

**Reply by FLIGHTLAMA** [reply]
Hi,
no problem you can close the idea. Thanks for looking into that.

However, I think it would be good at least to give a hint/modify the documentation (see snippet below)(https://docs.sqlalchemy.org/en/14/core/operators.html)
Because the query you get from this code with the default postgres dialect does not match the code in the example.
(VALUES is included in the example but the actual query executed on the db does not contain the VALUES  in the query).


**Snippet from the docs:**
```
Tuple IN Expressions
Comparison of tuples to tuples is common with IN, as among other use cases accommodates for the case when matching rows to a set of potential composite primary key values. The tuple_() construct provides the basic building block for tuple comparisons. The Tuple.in_() operator then receives a list of tuples:

>>> from sqlalchemy import tuple_
>>> tup = tuple_(column('x', Integer), column('y', Integer))
>>> expr = tup.in_([(1, 2), (3, 4)])
>>> print(expr)
(x, y) IN ([POSTCOMPILE_param_1])
To illustrate the parameters rendered:

>>> tup = tuple_(User.id, Address.id)
>>> stmt = select(User.name).join(Address).where(tup.in_([(1, 1), (2, 2)]))
>>> conn.execute(stmt).all()
SELECT user_account.name
FROM user_account JOIN address ON user_account.id = address.user_id
WHERE (user_account.id, address.id) IN (VALUES (?, ?), (?, ?))
[...] (1, 1, 2, 2)
[('spongebob',), ('sandy',)]
```

**Reply by zzzeek** [reply]
what's the reason we can't change it to use VALUES in all cases?   if it has better query plans.  or is that conditional based on if the tuples contain only literal values?     an enhancement can be made to achieve this, if the usage of VALUES depends on no SQL expressions embedded.

**Reply by zzzeek** [reply]
things like https://dba.stackexchange.com/questions/245394/efficient-select-for-array-of-tuples-in-postgresql suggest there are tradeoffs in both directions.   so, perhaps getting the existing values() construct to work here is best (it doesn't work right now?)

**Reply by zzzeek** [reply]
doesn't work right now, it's trying to treat it like a table.  OK

**Reply by zzzeek** [reply]
> Hi,
> no problem you can close the idea. Thanks for looking into that.
> 
> However, I think it would be good at least to give a hint/modify the documentation (see snippet below)(https://docs.sqlalchemy.org/en/14/core/operators.html)
> Because the query you get from this code with the default postgres dialect does not match the code in the example.
> (VALUES is included in the example but the actual query executed on the db does not contain the VALUES in the query).

well that's the SQLite dialect.    all the SQL in those tutorials will look different for different backends.

if we add an option to force the use of "values" that would be added in this section as a "Tip:" box.

**Reply by CaselIT** [reply]
>  (it doesn't work right now?)

@zzzeek see #6289
